Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1332 @@
+
+```hs
+>>> parseTest atom "false"
+SBool False
+
+>>> parseTest atom "hey"
+1:1:
+  |
+1 | hey
+  | ^^^
+unexpected "hey"
+expecting "false" or "true"
+```
+
+### Integers
+
+To parse integers, one strategy is to parse as many digits (chars whose value is between `'0'` and `'9'`) as possible. This will give us a list of characters, which we can then `read`,
+
+For this parser, we use the `some` function, which tries to run some given parser **1 or more times**. `numberChar` is exported from `Text.Megaparsec.Char` and parses any character that matches a digit. In regex, this is equivalent to `[0-9]+`. This will return us a `[Char]` (aka `String`), which we can `read` to make it into an `Integer`.
+
+```hs
+integer :: Parser Integer
+integer = read <$> some numberChar
+```
+
+We now change our `atom` to also include integers.
+
+```hs
+atom :: Parser SExp
+atom = choice
+  [ SBool <$> bool
+  , SInteger <$> integer
+  ]
+```
+
+Let's also test it in GHCi:
+
+```hs
+>>> parseTest atom "42"
+SInteger 42
+
+>>> parseTest atom "hey"
+1:1:
+  |
+1 | hey
+  | ^^^
+unexpected "hey"
+expecting "false", "true", or numeric character
+```
+
+Although the error message is good, wouldn't it better if it said "integer" instead of "numeric character"? For this, we can use **labels**, which allows us to give a name to a parser. To do this, we can simply change our parser like this:
+
+```hs
+integer :: Parser Integer
+integer = label "integer" $ read <$> some numberChar
+```
+
+And now the error message should be slightly improved:
+
+```hs
+>>> parseTest atom "hey"
+1:1:
+  |
+1 | hey
+  | ^^^
+unexpected "hey"
+expecting "false", "true" or integer
+```
+
+You can also use `label "boolean"` in our `bool` parser to change the message to `expecting boolean or integer`.
+
+**Exercise**: Can you change this parser so it also accepts negative integers?
+
+<details>
+  <summary>Solution</summary>
+One possible solution using `<|>`:
+
+```hs
+integer :: Parser Integer
+integer = label "integer" $ read <$> (some numberChar <|> ((:) <$> char '-' <*> some numberChar))
+```
+
+However, later in the tutorial, we will see an even better way that uses Megaparsec's own mechanisms.
+  <hr>
+</details>
+
+### Strings
+
+To parse a string, the idea is to find everything that is between double-quotes. This approach has a weakness, though: we won't be able to parse escaped quotes. Later in this tutorial, we will present a solution to escaped characters.
+
+To achieve this, we will use two new functions: `between` and `takeWhileP`. Using `between open close x` is equivalent to `open *> x <* close`, while `takeWhileP` will read as many characters as possible that satisfy some predicate.
+
+Note that `takeWhileP` also accepts, optionally, a label for each character that it tries to read. For now, we will leave it as `Nothing`.
+
+```hs
+str :: Parser String
+str = label "string" $ between (char '"') (char '"') (takeWhileP Nothing (/= '"'))
```

Comment:
```hs
str = do
  char '"'
  charLiteral `manyTill` char '"'
```

Reply:
This is how `str` is presented later in the article. But I added another point explaining this part a bit more.